### PR TITLE
fix!: realtime v2

### PR DESCRIPTION
**Pull request by grdsdev**
## What kind of change does this PR introduce?

**THIS PR HAS BREAKING CHANGES!**

This PR provides several fixes for realtime, including:
* Broadcast
* Postgres Changes
* Presence

## Breaking Changes
TBD

### CODE REVIEW

File: realtime/channel.py [hunk 1]
```diff
@@ -3,69 +3,170 @@
...
+            return
+
+        self.timeout_task.cancel()
+        self.timeout_task = None
+
+    def _match_receive(self, status: str, response: Any):
+        for hook in self.rec_hooks:
+            if hook.status == status:
+                # FIXME: callback can be a coroutine, how to handle that?
+                hook.callback(response)
```
Comment (grdsdev) [inline realtime/channel.py:145]: python experts: how to handle this?

The way it is, there are warnings saying the coroutine was never awaited.

**Comment by juancarlospaco** [comment]
> I had to use **kargs on the callbacks

I think it is Ok there in this context,
You can also use `*args` and use the arguments passed by their names.

File: realtime/channel.py (hunk 1 above)
Comment (juancarlospaco) [reply realtime/channel.py:145]: Just `await hook.callback(response)` ?

For more precise checking:

```python
import asyncio

def check_coroutine(callback):
    if asyncio.iscoroutine(callback):
        print("COROUTINE")
    else:
        print("NOT COROUTINE")

async def coroutine_function():
    pass

def normal_function():
    pass

check_coroutine(coroutine_function)
check_coroutine(normal_function)
```

Maybe code sample is better to understand.

Comment (grdsdev) [reply realtime/channel.py:145]: if I just `await hook.callback(response)` but it isn't a coroutine at runtime, wouldn't it fail/warn too?

Comment (juancarlospaco) [reply realtime/channel.py:145]: Edited it, see code sample.

File: realtime/presence.py [hunk 2]
```diff
@@ -2,18 +2,70 @@
  Defines the RealtimePresence class and its dependencies.
 """
 
-from typing import Any, Callable, Dict, List, Optional
+from enum import StrEnum
```
Comment (juancarlospaco) [inline realtime/presence.py:5]: `enum.StrEnum` was added in Python `3.11.0`, [see this link last line of that paragraph.](https://devdocs.io/python~3.12/library/enum#enum.StrEnum)

Before that you gotta use something like:

```python
import enum

class MyEnum(enum.Enum, str):
  KEY0 = "VALUE0"
  KEY1 = "VALUE1"

```

File: realtime/connection.py [hunk 3]
```diff
@@ -37,41 +46,37 @@ def __init__(
...
+                    Also accepts default Supabase URL: `http://` or `https://`.
+        :param token: Authentication token for the WebSocket connection.
+        :param auto_reconnect: If True, automatically attempt to reconnect on disconnection. Defaults to False.
+        :param params: Optional parameters for the connection. Defaults to an empty dictionary.
+        :param hb_interval: Interval (in seconds) for sending heartbeat messages to keep the connection alive. Defaults to 30.
+        :param max_retries: Maximum number of reconnection attempts. Defaults to 5.
+        :param initial_backoff: Initial backoff time (in seconds) for reconnection attempts. Defaults to 1.0.
         """
-        self.url = f"{re.sub(r'https?://', 'wss://', url, flags=re.IGNORECASE)}/realtime/v1/websocket?apikey={token}"
+        self.url = f"{re.sub(r'https://', 'wss://', re.sub(r'http://', 'ws://', url, flags=re.IGNORECASE), flags=re.IGNORECASE)}/websocket?apikey={token}"
```
Comment (juancarlospaco) [inline realtime/connection.py:65]:
```suggestion
        self.url = f"{re.sub(r'https://', 'wss://', re.sub(r'http://', 'ws://', url, flags=re.IGNORECASE), flags=re.IGNORECASE)}/realtime/v1/websocket?apikey={token}"
```

It is missing the `"/v1/"` thing as it had previously ? 🤔

Comment (grdsdev) [reply realtime/connection.py:65]: `supabase-py` should create the socket passing the url with `/realtime/v1` already, need to change it on initialization.

**Comment by grdsdev** [comment]
> > I had to use **kargs on the callbacks
> 
> I think it is Ok there in this context, You can also use `*args` and use the arguments passed by their names.

Thanks,  `*args` makes usage better I think.

File: Makefile [hunk 4]
```diff
@@ -13,7 +13,7 @@ tests_pre_commit:
 run_tests: tests
 
 tests_only:
-	poetry run pytest --cov=./ --cov-report=xml --cov-report=html -vv
```
Comment (grdsdev) [inline Makefile:16]: it was failing on CI, don't know why, need to check it after.

File: tests/test_connection.py [hunk 5]
```diff
@@ -1,17 +1,238 @@
...
+    assert received_events[0]["payload"]["message"] == "Event 1"
+    assert received_events[1]["payload"]["message"] == "Event 2"
+    assert received_events[2]["payload"]["message"] == "Event 3"
+
+    await socket.close()
+    listen_task.cancel()
+
+
+@pytest.mark.asyncio
+async def test_postgrest_changes(socket: Socket):
```
[the remark this answers is not in the document]
Comment (grdsdev) [reply tests/test_connection.py]: I doubled timeout from 5 to 10 seconds, still didn't work, but I realize that even locally, it always failed the first time it ran, so that is interesting, I've changed CI to run tests twice with:

```
make run_tests || true
make run_tests
```

And now it is passing, @filipecabaco any clue on why that happens?

[the remark this answers is not in the document]
Comment (filipecabaco) [reply tests/test_connection.py]: no clue so weird :| ... do you see anything outstanding in logs locally?

File: usage.py [hunk 6]
```diff
@@ -1,23 +1,61 @@
...
-def callback1(payload):
-    print("Callback 1: ", payload)
 
+async def main():
+    REALTIME_URL = "ws://localhost:4000/websocket"
+    API_KEY = "1234567890"
 
-def callback2(payload):
-    print("Callback 2: ", payload)
+    socket = RealtimeClient(REALTIME_URL, API_KEY)
```
Comment (filipecabaco) [inline usage.py:12]: how will this look when we use in `supabase-py` ?

**Review by filipecabaco — APPROVED** [review]
not as versed in Python as I wished but the functionality and API seems sane and close to other libs 🔥 

I also have a question on what will the `supabase-py` usage look like but probably that will be handled in a PR there.

File: usage.py (hunk 6 above)
Comment (grdsdev) [reply usage.py:12]: just like js, SupabaseClient gives access to the realtime client instance, but also exposes convenience methods for managing channels, example:

```python
supabase = create_client(...)

supabase.realtime # for full access

channel = supabase.channel("topic", params={"config": {"broadcast": {"self": True}}})
all_channels = supabase.get_all_channels()
```

We still need to implement this on `supabase-py`, but that is the ideia, if you have any recommendation, it is welcomed.

Comment (filipecabaco) [reply usage.py:12]: perfect image

**Review by juancarlospaco — APPROVED** [review]
👍

**Review by juancarlospaco — APPROVED** [review]
LGTM

File: pyproject.toml [hunk 7]
```diff
@@ -11,20 +11,24 @@ readme = "README.md"
...
 pytest = "^8.3.1"
 python-dotenv = "^1.0.1"
+pytest-asyncio = "^0.23.8"
 
 [tool.poetry.group.dev.dependencies]
 black = ">=23.11,<25.0"
 isort = "^5.12.0"
 pre-commit = "^3.5.0"
 pytest-cov = "^5.0.0"
+unasync-cli = { git = "https://github.com/supabase-community/unasync-cli.git", branch = "main" }
```
Comment (grdsdev) [inline pyproject.toml:31]: TODO: fix before merge

File: README.md [hunk 8]
```diff
@@ -1,73 +1,203 @@
...
+
+channel.subscribe(
+    lambda status, err: status == RealtimeSubscribeStates.SUBSCRIBED
+    and print("Ready to receive database changes!")
+)
 ```
 
+## Get All Channels
 
+You can see all the channels that your client has instantiatied.
```
Comment (juancarlospaco) [inline README.md:159]:
```suggestion
You can see all the channels that your client has instantiated.
```

File: Makefile (hunk 4 above)
Comment (J0) [reply Makefile:16]: Maybe coverage requirements?  As an aside probably can consider moving towards using [poetry scripts](https://python-poetry.org/docs/cli/#run) in the future when time frees up

File: README.md [hunk 9]
```diff
@@ -1,73 +1,203 @@
...
+await channel.subscribe(_on_subscribe)
+
+# Listen for all incoming events, often the last thing you want to do.
+await client.listen()
 ```
 
-## Quick Start 
+### Notes:
+
+- `REALTIME_URL` is `'ws://localhost:4000/socket'` when developing locally and `'wss://<project_ref>.supabase.co/realtime/v1'` when connecting to your Supabase project.
```
Comment (J0) [inline README.md:69]: /nit very minor point - would personally typically copy paste in browser so wondering  if we should consider removing the quotes so folks can directly copy paste

```suggestion
- `REALTIME_URL` is `ws://localhost:4000/socket` when developing locally and `wss://<project_ref>.supabase.co/realtime/v1` when connecting to your Supabase project.
```

File: README.md [hunk 10]
```diff
@@ -1,73 +1,203 @@
...
+    {
+        "config": {
+            "presence": {
+                "key": ""
+            }
+        }
+    }
+)
+
+channel.on_presence_sync(lambda: print("Online users: ", channel.presenceState()))
```
Comment (J0) [inline README.md:112]: Can't seem to find `presenceState`. Have we implemented it yet?

```suggestion
channel.on_presence_sync(lambda: print("Online users: ", channel.presenceState()))
```

File: example/app.py [hunk 11]
```diff
@@ -1,53 +1,141 @@
...
+    assert received_events[0]["payload"]["message"] == "Event 1"
+    assert received_events[1]["payload"]["message"] == "Event 2"
+    assert received_events[2]["payload"]["message"] == "Event 3"
+
+
+async def test_postgres_changes(socket: AsyncRealtimeClient):
+    await socket.connect()
+
+    await socket.set_auth(
+        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJhdXRoZW50aWNhdGVkIiwiZXhwIjoxNzIzNzUzNjI3LCJpYXQiOjE3MjM3NTAwMjcsImlzcyI6Imh0dHA6Ly8xMjcuMC4wLjE6NTQzMjEvYXV0aC92MSIsInN1YiI6ImYwNDBmNmY4LTA2YTUtNDA1My04YzQzLTcyMjBlZWMzZDMyZSIsImVtYWlsIjoidGVzdEBtYWlsLmNvbSIsInBob25lIjoiIiwiYXBwX21ldGFkYXRhIjp7InByb3ZpZGVyIjoiZW1haWwiLCJwcm92aWRlcnMiOlsiZW1haWwiXX0sInVzZXJfbWV0YWRhdGEiOnt9LCJyb2xlIjoiYXV0aGVudGljYXRlZCIsImFhbCI6ImFhbDEiLCJhbXIiOlt7Im1ldGhvZCI6InBhc3N3b3JkIiwidGltZXN0YW1wIjoxNzIzNzUwMDI3fV0sInNlc3Npb25faWQiOiJkMzA2OTc0Ni1iOTE5LTRkOTgtOWQ2Mi01MmRhYTZmMmIyZTgiLCJpc19hbm9ueW1vdXMiOmZhbHNlfQ.k-lLY0ogNZKEsvZNXP5xgi4hgBcRh0IbQtkc-Y_QXcs"
```
Comment (J0) [inline example/app.py:70]: Probably doesn't affect the test but can consider setting a later expiry like below as the token above will expire today 

```suggestion    "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhdWQiOiJhdXRoZW50aWNhdGVkIiwiZXhwIjoyMDIzNzUzNjI3LCJpYXQiOjE3MjM3NTAwMjcsImlzcyI6Imh0dHA6Ly8xMjcuMC4wLjE6NTQzMjEvYXV0aC92MSIsInN1YiI6ImYwNDBmNmY4LTA2YTUtNDA1My04YzQzLTcyMjBlZWMzZDMyZSIsImVtYWlsIjoidGVzdEBtYWlsLmNvbSIsInBob25lIjoiIiwiYXBwX21ldGFkYXRhIjp7InByb3ZpZGVyIjoiZW1haWwiLCJwcm92aWRlcnMiOlsiZW1haWwiXX0sInVzZXJfbWV0YWRhdGEiOnt9LCJyb2xlIjoiYXV0aGVudGljYXRlZCIsImFhbCI6ImFhbDEiLCJhbXIiOlt7Im1ldGhvZCI6InBhc3N3b3JkIiwidGltZXN0YW1wIjoxNzIzNzUwMDI3fV0sInNlc3Npb25faWQiOiJkMzA2OTc0Ni1iOTE5LTRkOTgtOWQ2Mi01MmRhYTZmMmIyZTgiLCJpc19hbm9ueW1vdXMiOmZhbHNlfQ.RL9gim2Zjcy8K87FLFQaNV5XQLT1AdoXxlPrg0oNpEc"
```

File: example/app.py [hunk 12]
```diff
@@ -1,53 +1,141 @@
...
-
-    # Provide the required channel_params argument
-    socket_channel = ss.set_channel(channel.topic, channel_params={})
-    socket_channel.on_broadcast("test", broadcast_callback).subscribe()
-    socket_channel.send_broadcast("test", {"message": "Hello from the other side"})
-    time.sleep(6)
+    socket = AsyncRealtimeClient(f"{URL}/realtime/v1", JWT, auto_reconnect=True)
+    await socket.connect()
+
+    # await test_broadcast_events(socket)
```
Comment (J0) [inline example/app.py:136]:
```suggestion
```

File: pyproject.toml [hunk 13]
```diff
@@ -11,20 +11,24 @@ readme = "README.md"
 repository = "https://github.com/supabase-community/realtime-py"
 
 [tool.poetry.dependencies]
-python = "^3.8"
+python = "^3.8.18"
```
Comment (J0) [inline pyproject.toml:14]: Consider moving to Python 3.9 as it looks like we might have [deprecated support for 3.8
](https://github.com/supabase/realtime-py/pull/182) while the PR was in progress

```suggestion
python = "^3.9.8"
```

File: realtime/_async/client.py [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+    DEFAULT_TIMEOUT,
+    PHOENIX_CHANNEL,
+    Callback,
+    ChannelEvents,
+    T_ParamSpec,
+    T_Retval,
+)
+from .channel import AsyncRealtimeChannel, RealtimeChannelOptions
+
+# logging.basicConfig(
```
Comment (J0) [inline realtime/_async/client.py:23]:
```suggestion
```

**Review by J0 — APPROVED** [review]
Nice work

File: README.md (hunk 10 above)
Comment (grdsdev) [reply README.md]: nice catch, thanks

File: pyproject.toml [hunk 15]
```diff
@@ -11,20 +11,24 @@ readme = "README.md"
 repository = "https://github.com/supabase-community/realtime-py"
 
 [tool.poetry.dependencies]
-python = "^3.8"
+python = "^3.9"
 websockets = ">=11,<13"
 python-dateutil = "^2.8.1"
 typing-extensions = "^4.12.2"
+asyncio = "^3.4.3"
```
Comment (timkpaine) [inline pyproject.toml:18]: why is this added? this is not a relevant package since python 3.3?
